### ISSUE TITLE
Exception-based error reporting is noisy

### ISSUE THREAD

**Issue by yminsky** [janestreet/ppx_inline_test#26]
I'm working on the testing chapter of Real World OCaml, and it highlights that using `[%test_eq[` leads to rather noisy error messages, with mostly-nonsense stack traces.  Here's the example:

```ocaml
open! Base

let%test_unit "rev" =
  [%test_eq: int list] (List.rev [3;2;1]) [3;2;1]
```
And here's what the test output looks like.

```sh
  $ dune runtest
  File "test.ml", line 3, characters 0-71: rev threw
  (duniverse/ppx_assert.v0.13.0/runtime-lib/runtime.ml.E "comparison failed"
    ((1 2 3) vs (3 2 1) (Loc test.ml:4:13))).
    Raised at file "duniverse/ppx_assert.v0.13.0/runtime-lib/runtime.ml", line 28, characters 28-53
    Called from file "duniverse/ppx_inline_test.v0.13.1/runtime-lib/runtime.ml", line 502, characters 15-19
    Called from file "duniverse/ppx_inline_test.v0.13.1/runtime-lib/runtime.ml", line 343, characters 8-12
    Re-raised at file "duniverse/ppx_inline_test.v0.13.1/runtime-lib/runtime.ml", line 346, characters 6-13
    Called from file "duniverse/ppx_inline_test.v0.13.1/runtime-lib/runtime.ml", line 359, characters 15-52
    Called from file "duniverse/ppx_inline_test.v0.13.1/runtime-lib/runtime.ml", line 446, characters 52-83

  FAILED 1 / 1 tests
  [1]
```

Is this fixable?

**Comment by sliquister** [comment]
It's hard to line up the positions of these stack traces with the actual source code, but it should be fixable. I think it's a matter of capturing exceptions directly from the user code, whereas right now there are things like Exn.protect in between the user code and the try with.

**Comment by sliquister** [comment]
I have a couple of changes under review to go from:

```
File "source.ml", line 41, characters 0-75: rev threw
(runtime.ml.E "comparison failed"
  ((1 2 3) vs (3 2 1) (Loc sourcelib/source.ml:42:13))).
  Raised at Ppx_assert_lib__Runtime.failwith in file "runtime.ml" (inlined), line 28, characters 28-53
  Called from Ppx_assert_lib__Runtime.fail_in_sexp_style in file "runtime.ml" (inlined), line 48, characters 2-23
  Called from Ppx_assert_lib__Runtime.test_eq_failed in file "runtime.ml", line 85, characters 2-107
  Called from Ppx_inline_test_lib__Runtime.test_unit.(fun) in file "runtime.ml", line 513, characters 15-19
  Called from Base__Exn.protectx in file "exn.ml" (inlined), line 66, characters 8-11
  Called from Base__Exn.protect in file "exn.ml", line 77, characters 26-49
  Re-raised at Base__Exn.protectx in file "exn.ml" (inlined), line 71, characters 4-114
  Called from Base__Exn.protect in file "exn.ml", line 77, characters 26-49
  Called from Ppx_inline_test_lib__Runtime.time_and_reset_random_seeds in file "runtime.ml" (inlined), line 365, characters 15-52
  Called from Ppx_inline_test_lib__Runtime.test in file "runtime.ml", line 460, characters 53-84
```

to:

```
File "source.ml", line 41, characters 0-75: rev threw
(runtime.ml.E "comparison failed"
  ((1 2 3) vs (3 2 1) (Loc sourcelib/source.ml:42:13))).
  Called from Ppx_assert_lib__Runtime.test_eq in file "runtime.ml", line 85, characters 2-107
  Called from Sourcelib__source.(fun) in file "source.ml", line 42, characters 13-21
```

which I think as it good as it can get.

**Comment by yminsky** [comment]
That looks amazing. Thanks!

**Comment by aalekseyev** [comment]
This change landed in 4ad759e0d454fba07007901f5221214cb4cd8cd5.
